src/imperative/imperative_utils.h

```diff
@@ -379,7 +379,8 @@ inline void PushFCompute(const FCompute& fn,
                              &input_blobs, &output_blobs, &pre_temp_src, &pre_temp_dst,
                              &post_temp_src, &post_temp_dst, &in_temp_idx_map, mutate_idx);
       // setup context
-      OpContext opctx{is_train, rctx, engine::CallbackOnComplete(), requested};
+      bool need_grad = Imperative::Get()->is_recording();
```

Comment:
need_grad shouldn't be get from the worker thread. It should be set outside similar to is_train